File: workers/research/src/black_book_research/relevance/public.py
"""Public relevance projection without numeric scores (BB-040)."""

from __future__ import annotations

import json
from dataclasses import asdict

from .types import PublicRelevanceExplanation, RelevanceAssessment
from .why import assert_explanation_has_no_numeric_score


def to_public_relevance_explanation(assessment: RelevanceAssessment) -> PublicRelevanceExplanation:
    assert_explanation_has_no_numeric_score(assessment.why_this_appears)
    return PublicRelevanceExplanation(
        why_this_appears=assessment.why_this_appears,
        decision=assessment.decision,
    )


def assert_public_relevance_has_no_score(
    public_explanation: PublicRelevanceExplanation,
    assessment: RelevanceAssessment,
) -> None:
    serialized = json.dumps(asdict(public_explanation))
    if str(assessment.composite_score) in serialized:
        raise ValueError("Public relevance projection must not expose composite score.")
    assert_explanation_has_no_numeric_score(public_explanation.why_this_appears)

File: packages/domain/src/relevance/public.ts
/**
 * Public relevance projection hides numeric scores.
 */
import type { PublicRelevanceExplanation, RelevanceAssessment } from './types.js';
import { assertExplanationHasNoNumericScore } from './why.js';

export function toPublicRelevanceExplanation(
  assessment: RelevanceAssessment,
): PublicRelevanceExplanation {
  assertExplanationHasNoNumericScore(assessment.whyThisAppears);
  return {
    whyThisAppears: assessment.whyThisAppears,
    decision: assessment.decision,
  };
}

export function assertPublicRelevanceHasNoScore(
  publicExplanation: PublicRelevanceExplanation,
  assessment: RelevanceAssessment,
): void {
  const serialized = JSON.stringify(publicExplanation);
  if (serialized.includes(String(assessment.compositeScore))) {
    throw new Error('Public relevance projection must not expose composite score.');
  }
  if ('compositeScore' in publicExplanation) {
    throw new Error('Public relevance projection must not include compositeScore field.');
  }
  if ('featureValues' in publicExplanation) {
    throw new Error('Public relevance projection must not include featureValues field.');
  }
  assertExplanationHasNoNumericScore(publicExplanation.whyThisAppears);
}
